### PR TITLE
armv7m: Enable MPU

### PR DESCRIPTION
**Pull request by agkaminski**
## Description

CI on armv7m targets **will not pass**, as platform configuration became invalid (apps need to be provided with multiple allowed maps, i/o maps are not defined)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (refactoring, style fixes, git/CI config, submodule management, no code logic changes)

## How Has This Been Tested?

- [x] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: (stm32l4x6, imxrt1064).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [x] This PR needs additional PRs to work (https://github.com/phoenix-rtos/phoenix-rtos-project/pull/1076).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: hal/armv7m/pmap.c [hunk 1]
```diff
@@ -36,9 +49,53 @@ addr_t pmap_destroy(pmap_t *pmap, int *i)
...
+}
+
+
 void pmap_switch(pmap_t *pmap)
 {
-	return;
+	unsigned int i;
+
+	if (pmap != NULL) {
+		for (i = 0; i < syspage->hs.mpu.allocCnt; ++i) {
```
Comment (kemonats) [inline hal/armv7m/pmap.c:84]: What do you think about this solution, faster and smaller result code.
```suggestion
	unsigned int i;
	const unsigned int allocCnt = syspage->hs.mpu.allocCnt;

	if (pmap != NULL) {
		for (i = 0; i < allocCnt; ++i) {
```

Comment (agkaminski) [reply hal/armv7m/pmap.c:84]: Thanks, will use, for now I still need to debug some problems.
Btw, nice to see you back ;)

File: proc/msg-nommu.c [hunk 2]
```diff
@@ -129,19 +145,50 @@ int proc_recv(u32 port, msg_t *msg, msg_rid_t *rid)
...
+			return -ENOMEM;
+		}
+		hal_memcpy(idata, kmsg->msg->i.data, kmsg->msg->i.size);
+		kmsg->imapped = idata;
+		msg->i.data = idata;
+	}
+
+	if ((kmsg->msg->o.data != NULL) && (kmsg->msg->o.size != 0) && (current->process != NULL) &&
+			(pmap_isAllowed(current->process->pmapp, kmsg->msg->o.data, kmsg->msg->o.size) == 0)) {
+		msg->o.data = vm_mmap(current->process->mapp, NULL, NULL, round_page(kmsg->msg->o.size),
```
Comment (lukileczo) [inline proc/msg-nommu.c:176]: I'd add some space here
```suggestion

		msg->o.data = vm_mmap(current->process->mapp, NULL, NULL, round_page(kmsg->msg->o.size),
```

File: proc/msg-nommu.c [hunk 3]
```diff
@@ -129,19 +145,50 @@ int proc_recv(u32 port, msg_t *msg, msg_rid_t *rid)
...
 	*rid = lib_idtreeId(&kmsg->idlinkage);
 
 	hal_memcpy(msg, kmsg->msg, sizeof(*msg));
 
+	kmsg->imapped = NULL;
+	kmsg->omapped = NULL;
+
+	if ((kmsg->msg->i.data != NULL) && (kmsg->msg->i.size != 0) && (current->process != NULL) &&
+			(pmap_isAllowed(current->process->pmapp, kmsg->msg->i.data, kmsg->msg->i.size) == 0)) {
+		idata = vm_mmap(current->process->mapp, NULL, NULL, round_page(kmsg->msg->i.size),
```
Comment (lukileczo) [inline proc/msg-nommu.c:161]:
```suggestion

		idata = vm_mmap(current->process->mapp, NULL, NULL, round_page(kmsg->msg->i.size),
```

File: proc/msg-nommu.c (hunk 2 above)
Comment (agkaminski) [reply proc/msg-nommu.c:178]: Done

File: proc/msg-nommu.c [hunk 4]
```diff
@@ -129,19 +145,52 @@ int proc_recv(u32 port, msg_t *msg, msg_rid_t *rid)
...
+			PROT_READ | PROT_WRITE | PROT_USER, NULL, -1, MAP_ANONYMOUS);
+		if (msg->o.data == NULL) {
+			if (idata != NULL) {
+				vm_munmap(current->process->mapp, idata, round_page(kmsg->msg->i.size));
+			}
+
+			/* Free RID */
+			(void)proc_portRidGet(p, *rid);
+			proc_msgReject(kmsg, p);
+			port_put(p, 0);
```
Comment (lukileczo) [inline proc/msg-nommu.c:188]: You call `port_put` twice here

Comment (agkaminski) [reply proc/msg-nommu.c]: Good catch, fixed, thanks

File: proc/process.c [hunk 5]
```diff
@@ -1096,15 +1098,31 @@ static void process_exec(thread_t *current, process_spawn_t *spawn)
...
+		}
+	}
 #endif
 
 	pmap_switch(current->process->pmapp);
 
-	err = process_load(current->process, spawn->object, spawn->offset, spawn->size, &stack, &entry);
+	if (err <= 0) {
+		err = process_load(current->process, spawn->object, spawn->offset, spawn->size, &stack, &entry);
+	}
```
Comment (lukileczo) [inline proc/process.c:1124]: Oh, and one more thing - from clang diagnostic, `entry` is used not initialized when `err > 0`. This won't happen with current code path as `pmap_addMap` returns `<= 0` so maybe it's ok to remove this `if`?

Comment (agkaminski) [reply proc/process.c]: Ugh, this condition was a major brain-fart, fixed. I also fixed this potential usage of not initialized `entry`

**Review by lukileczo — DISMISSED** [review]
LGTM

File: proc/process.c [hunk 6]
```diff
@@ -1096,15 +1098,31 @@ static void process_exec(thread_t *current, process_spawn_t *spawn)
 #ifndef NOMMU
 	vm_mapCreate(&current->process->map, (void *)(VADDR_MIN + SIZE_PAGE), (void *)VADDR_USR_MAX);
 	proc_changeMap(current->process, &current->process->map, NULL, &current->process->map.pmap);
+	(void)i;
 #else
+	pmap_create(&current->process->map.pmap, NULL, NULL, NULL);
```
Comment (Darchiv) [inline proc/process.c:1103]: Can `pmap_create()` fail on any arch? Or do we assume that armv7m's implementation always succeeds?

File: proc/process.c [hunk 7]
```diff
@@ -1096,15 +1098,31 @@ static void process_exec(thread_t *current, process_spawn_t *spawn)
...
+		for (i = 0; (i < spawn->prog->dmapSz) && (err >= 0); ++i) {
+			err = pmap_addMap(current->process->pmapp, spawn->prog->dmaps[i]);
+		}
+	}
 #endif
 
 	pmap_switch(current->process->pmapp);
 
-	err = process_load(current->process, spawn->object, spawn->offset, spawn->size, &stack, &entry);
+	if (err == 0) {
```
Comment (Darchiv) [inline proc/process.c:1122]: What if `pmap_addMap()` will return >=0 values in the future? (I assume this is what you mean by `(err >= 0)` in for loops' conditions above.) Is this condition still valid? There is also a second `err == 0`, and also `if (err != 0)` which is reachable with `pmap_addMap()`'s value

File: proc/process.c (hunk 6 above)
Comment (agkaminski) [reply proc/process.c:1103]: Imho we can change the HAL API, so pmap_create has no return code - none of the architectures has any error path anyway

File: proc/process.c (hunk 7 above)
Comment (agkaminski) [reply proc/process.c:1122]: pmap_addMap does not return values > 0

File: hal/armv7m/pmap.c [hunk 8]
```diff
@@ -93,11 +166,74 @@ int pmap_segment(unsigned int i, void **vaddr, size_t *size, int *prot, void **t
...
+	 * map. Possible fix - place return to the user code in the separate
+	 * region and allow this region instead. */
+
+	/* Find kernel code region */
+	__asm__ volatile ("\tmov %0, pc;" : "=r" (pc));
+	ikmap = syspage_mapAddrResolve(pc);
+	if (ikmap == NULL) {
+		hal_consolePrint(ATTR_BOLD, "pmap: Kernel code map not found. Bad system config\n");
+		for (;;) {
+		}
```
Comment (gerard5) [inline hal/armv7m/pmap.c:228]:
```suggestion
		for (;;) {
			hal_cpuHalt();
		}
```

File: hal/armv7m/pmap.c [hunk 9]
```diff
@@ -93,11 +166,74 @@ int pmap_segment(unsigned int i, void **vaddr, size_t *size, int *prot, void **t
...
+	}
+
+	ikregion = pmap_map2region(ikmap->id);
+	if (ikregion < 0) {
+		hal_consolePrint(ATTR_BOLD, "pmap: Kernel code map has no assigned region. Bad system config\n");
+		for (;;) {
+		}
```
Comment (gerard5) [inline hal/armv7m/pmap.c:235]:
```suggestion
		for (;;) {
			hal_cpuHalt();
		}
```

File: proc/process.c (hunk 7 above)
Comment (agkaminski) [reply proc/process.c:1122]: Changed condition check in `for`s above for consistency

**Review by Darchiv — APPROVED** [review]

**Pull request merged**